// lit-element-effect/dist/decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withEffect = exports.decorate = exports.asUpdateableLitElement = void 0;
function asUpdateableLitElement(element) {
    if (!element.dispatchEvent || !element.requestUpdate)
        throw new Error("Element missing required functions (dispatchEvent/requestUpdate)");
    return element;
}
exports.asUpdateableLitElement = asUpdateableLitElement;
const reservedKeyword = "__registered_effects";
function decorate(litElement) {
    const decoratedLitElement = litElement;
    if (decoratedLitElement[reservedKeyword])
        return decoratedLitElement;
    const updateableLitLikeElement = asUpdateableLitElement(litElement);
    const oldUpdated = updateableLitLikeElement.updated;
    decoratedLitElement[reservedKeyword] = {
        index: 0,
        count: 0,
        effects: [],
    };
    updateableLitLikeElement.updated = (args) => {
        decoratedLitElement[reservedKeyword].index = 0;
        return oldUpdated(args);
    };
    return decoratedLitElement;
}
exports.decorate = decorate;
function withEffect(litElement, effect) {
    const decoratedLitElement = decorate(litElement);
    const { index, count } = decoratedLitElement[reservedKeyword];
    if (index === count) {
        decoratedLitElement[reservedKeyword].index++;
        decoratedLitElement[reservedKeyword].count++;
        decoratedLitElement[reservedKeyword].effects.push(effect);
        return effect;
    }
    decoratedLitElement[reservedKeyword].index++;
    return decoratedLitElement[reservedKeyword].effects[index];
}
exports.withEffect = withEffect;
